### PR TITLE
All Storage samples are now up to standard.

### PR DESCRIPTION
**Pull request by jmdobry**
- [x] - All samples use `yargs` for CLI
- [x] - All samples follow sample template pattern
- [x] - All tests follow same conventions

### CODE REVIEW

File: storage/encryption.js [hunk 1]
```diff
@@ -55,35 +55,26 @@ function generateEncryptionKey () {
...
   };
 
-  bucket.upload(srcFileName, options, function (err, file) {
+  bucket.upload(options.srcFile, config, function (err, file) {
     if (err) {
       return callback(err);
     }
 
-    console.log('Uploaded encrypted file: %s', destFileName);
+    console.log('Uploaded gs://%s/%s', options.bucket, options.destFile);
```
Comment (ace-n) [inline storage/encryption.js]: Should we make note of the uploaded file being encrypted?

File: storage/files.js [hunk 2]
```diff
@@ -72,326 +68,281 @@ function listFiles (name, callback) {
...
 
-  var bucket = storage.bucket(name);
-  var options = {
-    prefix: prefix
+  var config = {
+    prefix: options.prefix
   };
-  if (delimiter && typeof delimiter === 'string') {
-    options.delimiter = delimiter;
+  if (options.delimiter && typeof options.delimiter === 'string') {
```
Comment (ace-n) [inline storage/files.js]: Should we be checking types here? If `options.delimiter` isn't a string, I'm sure the user would want to be told that in a transparent way (e.g. an error message).

File: storage/files.js [hunk 3]
```diff
@@ -72,326 +68,281 @@ function listFiles (name, callback) {
...
+      description: 'Filter files by a prefix.'
+    },
+    delimiter: {
+      alias: 'd',
+      requiresArg: true,
+      type: 'string',
+      description: 'Specify a delimiter.'
+    }
+  }, function (options) {
+    if (options.prefix !== undefined) {
```
Comment (ace-n) [inline storage/files.js]: Does `if(options.prefix) { ... }` not work here?

File: storage/files.js [hunk 4]
```diff
@@ -72,326 +68,281 @@ function listFiles (name, callback) {
...
+  .command('delete <bucket> <file>', 'Delete a file from a bucket.', {}, function (options) {
+    program.deleteFile(options, console.log);
+  })
+  .command('getMetadata <bucket> <file>', 'Get metadata for a file in a bucket.', {}, function (options) {
+    program.getMetadata(options, console.log);
+  })
+  .command('makePublic <bucket> <file>', 'Make a file public in a bucket.', {}, function (options) {
+    program.makePublic(options, console.log);
+  })
+  .command('move <bucket> <srcFile> <destFile>', 'Rename a file in a bucket.', {}, function (options) {
```
Comment (ace-n) [inline storage/files.js]: "move" is a bit of a nebulous name for this command - could we call it "rename" within this file, or say something like "Move a file __within__ a bucket"?

File: storage/system-test/encryption.test.js [hunk 5]
```diff
@@ -54,24 +54,38 @@ describe('storage:encryption', function () {
...
     it('should upload a file', function (done) {
-      program.uploadEncryptedFile(bucketName, filePath, fileName, key, function (err, file) {
+      var options = {
+        bucket: bucketName,
+        srcFile: filePath,
+        destFile: fileName,
+        key: key
+      };
+
+      program.uploadEncryptedFile(options, function (err, file) {
```
Comment (ace-n) [inline storage/system-test/encryption.test.js]: What's the reasoning behind passing these in as an object?

File: storage/encryption.js (hunk 1 above)
Comment (jmdobry) [inline storage/encryption.js]: I think the method name (`uploadEncryptedFile`) and CLI (`node encryption upload ...`) cover it.

File: storage/files.js (hunk 2 above)
Comment (jmdobry) [inline storage/files.js]: Actually I should just remove this check.

File: storage/files.js (hunk 3 above)
Comment (jmdobry) [inline storage/files.js]: I was thinking `prefix` might be an empty string, but I think you're right. I'll change it.

File: storage/files.js (hunk 4 above)
Comment (jmdobry) [inline storage/files.js]: The Cloud Storage API (and `gsutil` CLI) mirror the S3 API and the also the linux `mv` command somewhat, e.g. `mv old_file.txt new_file.txt`. I think `move` is fine here. I'll clarify in the description though.

File: storage/system-test/encryption.test.js (hunk 5 above)
Comment (jmdobry) [inline storage/system-test/encryption.test.js]: All storage samples which before took more than one positional argument (besides the callback) have now collapsed them into a single `options` argument. This cleans up the function signature while also making it easier for the CLI to call the sample methods e.g. [`uploadEncryptedFile(options, callback)`](https://github.com/GoogleCloudPlatform/nodejs-docs-samples/blob/storage/storage/encryption.js#L58)

The JSDoc comment above each sample method very clearly documents the `options` argument and its properties.

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -27,6 +27,65 @@ accept your pull requests.
...
+// Instantiate a <API> client
+var <CLIENT> = <API_CLIENT>();
+// [END setup]
+
+// [START <REGION_TAG_NAME>]
+/**
+ * <DESCRIPTION>
+ *
+ * @param {<TYPE>} <NAME> <DESCRIPTION>.
+ * @param {function} cb The callback function.
```
Comment (jmdobry) [inline CONTRIBUTING.md]: `@param {<TYPE>} <NAME> <DESCRIPTION>.` can be repeated, which covers that.

File: storage/files.js [hunk 7]
```diff
@@ -72,326 +68,281 @@ function listFiles (name, callback) {
...
-  file.makePublic(function (err, apiResponse) {
+  file.makePublic(function (err) {
     if (err) {
       return callback(err);
     }
 
-    console.log('Made %s public!', fileName);
-    return callback(null, apiResponse);
+    console.log('Made gs://%s/%s public!', options.bucket, options.file);
+    return callback(null);
```
Comment (ace-n) [inline storage/files.js]: Is this null necessary? Can we just do `callback()`, or is that not clear enough?

File: storage/test/files.test.js [hunk 8]
```diff
@@ -56,425 +59,461 @@ function getSample () {
...
+      sample.program.copyFile(options, callback);
+
+      assert(callback.calledOnce, 'callback called once');
+      assert.equal(callback.firstCall.args.length, 1, 'callback received 1 argument');
+      assert(callback.firstCall.args[0], 'callback received error');
+      assert.equal(callback.firstCall.args[0].message, error.message, 'error has correct message');
     });
   });
+
   describe('main', function () {
```
Comment (ace-n) [inline storage/test/files.test.js]: I'm assuming we still want to keep this `main()` test around, even with `yargs` - or is that something we should discuss?

File: storage/test/transfer.test.js [hunk 9]
```diff
@@ -191,78 +172,55 @@ describe('storage:transfer', function () {
         var value = 'DISABLED';
```
Comment (ace-n) [inline storage/test/transfer.test.js]: What is the point of these variables if we're manipulating `options` and not them?

File: storage/test/transfer.test.js [hunk 10]
```diff
@@ -271,12 +229,19 @@ describe('storage:transfer', function () {
       });
 
       it('should handle patch error', function () {
+        var field = 'status';
```
Comment (ace-n) [inline storage/test/transfer.test.js]: Can these (field/value) be inserted directly into options?

File: storage/test/transfer.test.js [hunk 11]
```diff
@@ -191,78 +172,55 @@ describe('storage:transfer', function () {
...
-
-        assert(callback.calledOnce, 'callback called once');
-        assert.equal(callback.firstCall.args.length, 1, 'callback received 1 argument');
-        assert(callback.firstCall.args[0], 'callback received error');
-        assert.equal(callback.firstCall.args[0].message, '"value" is required!', 'error has correct message');
+        assert(console.log.calledWith('Updated transfer job: %s', sample.mocks.transferJob.name));
       });
 
       it('should handle auth error', function () {
+        var field = 'status';
```
Comment (ace-n) [inline storage/test/transfer.test.js]: Can these (field/value) be inserted directly into options?

File: storage/transfer.js [hunk 12]
```diff
@@ -51,20 +51,17 @@ function auth (callback) {
 /**
  * Review the transfer operations associated with a transfer job.
  *
```
Comment (ace-n) [inline storage/transfer.js]: extra line?

File: storage/system-test/encryption.test.js (hunk 5 above)
Comment (ace-n) [inline storage/system-test/encryption.test.js]: OK cool - thanks for clarifying!

File: storage/files.js (hunk 7 above)
Comment (jmdobry) [inline storage/files.js]: Not strictly necessary, but it's consistent with other samples in that `null` is passed as the first argument when there is no error. It's also part of the the [Node.js Errback Standard](http://fredkschott.com/post/2014/03/understanding-error-first-callbacks-in-node-js/#defining-an-error-first-callback:cb6a1cb7301215882ae9130ce938a4f7) to pass `null` as the first argument when there is no error.

File: storage/test/files.test.js (hunk 8 above)
Comment (jmdobry) [inline storage/test/files.test.js]: Yeah, there are a couple lines in each sample that get covered by having these `main` tests.

File: storage/test/transfer.test.js (hunk 9 above)
Comment (jmdobry) [inline storage/test/transfer.test.js]: You're right, fixed.

File: storage/test/transfer.test.js (hunk 10 above)
Comment (jmdobry) [inline storage/test/transfer.test.js]: Yes, fixed.

File: storage/test/transfer.test.js (hunk 11 above)
Comment (jmdobry) [inline storage/test/transfer.test.js]: Yes, fixed.

File: storage/transfer.js (hunk 12 above)
Comment (jmdobry) [inline storage/transfer.js]: As a practice, whenever I write JSDoc comments, I leave a line between the description and the tags. I think it helps "de-clutter" the comment block.

**Comment by ace-n** [comment]
...is there a reason test coverage dropped so much, or is `codecov.io` just being flaky?

(If it's just `codecov.io`, LGTM.)

**Comment by jmdobry** [comment]
There is a reason, I accidentally committed `.only` in one of the tests. Removing it. Will merge once tests pass.

**Comment by ace-n** [comment]
LGTM.

**Pull request merged**